modules/src/core/ics26_routing/context.rs

```diff
@@ -17,4 +34,187 @@ pub trait Ics26Context:
     + PortReader
     + Ics20Context
 {
+    type Router: Router;
+
+    fn router(&self) -> &Self::Router;
+
+    fn router_mut(&mut self) -> &mut Self::Router;
+}
+
+#[derive(Debug, PartialEq)]
+pub struct InvalidModuleId;
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct ModuleId(String);
+
+impl ModuleId {
+    pub fn new(s: Cow<'_, str>) -> Result<Self, InvalidModuleId> {
+        if !s.trim().is_empty() && s.chars().all(char::is_alphanumeric) {
+            Ok(Self(s.into_owned()))
+        } else {
+            Err(InvalidModuleId)
+        }
+    }
+}
+
+impl fmt::Display for ModuleId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl FromStr for ModuleId {
+    type Err = InvalidModuleId;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Self::new(Cow::Borrowed(s))
+    }
+}
+
+impl Borrow<str> for ModuleId {
+    fn borrow(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+/// Types implementing this trait are expected to implement `From<GenericAcknowledgement>`
+pub trait Acknowledgement: AsRef<[u8]> {
+    fn success(&self) -> bool;
+}
+
+pub type WriteFn = dyn FnOnce(&mut dyn Any);
+
+pub type DeferredWriteResult<T> = (Option<Box<T>>, Option<Box<WriteFn>>);
```

Comment:
Are all combinations possible here? ie. `(None, None)`, `(None, Some(_))`, `(Some(_), None)`, `(Some(_), Some(_))`

If not, then we should introduce an enum which captures only the possible combinations.

Reply:
Otherwise, if we keep the pair of `Option`s it would be nice to introduce an enum for each member of the pair instead of `Option` to provide a bit more semantic meaning here.

Reply:
This is a great idea! I think this can be better modelled as -> 
```rust
pub enum OnRecvPacketAck {
    Nil(Box<WriteFn>),
    Successful(Box<dyn Acknowledgement>, Box<WriteFn>),
    Failed(Box<dyn Acknowledgement>),
}
```

based on ibc-go code here -> 

https://github.com/cosmos/ibc-go/blob/ccc4cb804843f1a80acfb0d4dbf106d1ff2178bb/modules/core/keeper/msg_server.go#L410-L423

The reason I didn't do this before is that the `Acknowledgement` trait previously defined a `success()` method to determine if the ack was successful or not at runtime and I wasn't sure how to prevent users from creating `OnRecvPacketAck::Success` with a failed variant. I just realized that the `success()` method was only called by the handler so I removed the trait method and instead used the enum variant to communicate successful/failed acks.

Reply:
https://github.com/informalsystems/ibc-rs/pull/1838/commits/e148d5b91f0ee63c28e495716ae639c4d7072f7d